File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.java
import java.util.*;
class Solution {
    public int[] solution(String[] gems) {
        Set<String> set = new HashSet<>();
        Map<String, Integer> map = new HashMap<>();
        for(String gem : gems) {
            set.add(gem);
        }
        // 보석의 종류 수
        int start = 0;
        int end = 0;
        int minStart = 0;
        int minLength = gems.length;
        
        while(true) {
            // 모든 종류의 보석을 포함할 때까지
            if(map.size() < set.size()) {
                if(end == gems.length) break;
                map.put(gems[end], map.getOrDefault(gems[end], 0) + 1);
                end++;
            }
            else {
                // 모든 종류를 다 포함한다면 최소 구간 찾기
                if(end - start < minLength) {
                    minLength = end - start;
                    minStart = start;
                }
                
                map.put(gems[start], map.get(gems[start]) - 1);
                if(map.get(gems[start]) == 0)
                    map.remove(gems[start]);
                start++;
            }
        }
        
        int[] answer = {minStart + 1, minStart + minLength};
        return answer;
    }
}

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.py
def solution(gems):
    ans = []
    # 진열된 모든 종류의 보석을 적어도 1개 이상 포함하는 가장 짧은 구간
    s = set()
    for g in gems:
        s.add(g)
    # 목표 개수
    t = len(s)
    min_l = float('inf')
    l = 0
    d = dict()
    for r in range(len(gems)):
        gem = gems[r]
        d[gem] = d.get(gem,0) + 1
        while len(d.keys()) >= t:
            remove = gems[l]
            d[remove] = d.get(remove) - 1
            if d[remove] == 0:
                del d[remove]
            l += 1
            cur_l = r + 1 - l
            if cur_l < min_l:
                ans = [l,r+1]
                min_l = cur_l
    return ans